consensus/bor/bor.go

```diff
@@ -827,6 +828,15 @@ func (c *Bor) Prepare(chain consensus.ChainHeaderReader, header *types.Header) e
 	header.Time = parent.Time + CalcProducerDelay(number, succession, c.config)
 	if header.Time < uint64(time.Now().Unix()) {
 		header.Time = uint64(time.Now().Unix())
+	} else {
+		// For primary validators, wait until previous block time is completed before starting
+		// to produce next block
+		if succession == 0 {
+			startTime := time.Unix(int64(header.Time-c.config.CalculatePeriod(number)), 0)
+			if time.Now().Before(startTime) {
+				time.Sleep(time.Until(startTime))
+			}
+		}
```

Comment:
I believe we can avoid this section of code altogether. From what I understand, it's added to avoid bor to start building block before time. But I think that there's no harm in doing that because even if it starts and finishes before the actual block building window, the delay in seal function would avoid announcing block before `header.Time - 2s` timer hits. So, we'll anyways wait to announce block early. Wdyt?